go-controller/pkg/node/base_node_network_controller_dpu.go

```diff
@@ -14,139 +13,214 @@ import (
+// If true, return its dpuConnDetails, otherwise return nil
+func (bnnc *BaseNodeNetworkController) podReadyToAddDPU(pod *kapi.Pod, nadName string) *util.DPUConnectionDetails {
+	if bnnc.name != pod.Spec.NodeName {
+		klog.V(5).Infof("Pod %s/%s is not scheduled on this node %s", pod.Namespace, pod.Name, bnnc.name)
+		return nil
+	}
+
+	dpuCD, err := util.UnmarshalPodDPUConnDetails(pod.Annotations, nadName)
+	if err != nil {
+		if !util.IsAnnotationNotSetError(err) {
+			klog.Errorf("Failed to get dpu annotation for pod %s/%s nad %s: %v",
+				pod.Namespace, pod.Name, nadName, err)
+		} else {
+			klog.V(5).Infof("DPU connection details annotation still not found for %s/%s for NAD %s",
+				pod.Namespace, pod.Name, nadName)
+		}
+		return nil
+	}
+
+	return dpuCD
+}
+
+func (bnnc *BaseNodeNetworkController) addDPUPod4Nad(pod *kapi.Pod, dpuCD *util.DPUConnectionDetails, isOvnUpEnabled bool,
+	netName, nadName string, getter cni.PodInfoGetter) error {
+	podDesc := fmt.Sprintf("pod %s/%s for nad %s", pod.Namespace, pod.Name, nadName)
+	klog.Infof("Adding %s on DPU", podDesc)
+	podInterfaceInfo, err := cni.PodAnnotation2PodInfo(pod.Annotations, nil, isOvnUpEnabled,
+		string(pod.UID), "", nadName, netName, config.Default.MTU)
+	if err != nil {
+		klog.Errorf("Failed to get pod interface information of %s: %v. retrying", podDesc, err)
+		return err
+	}
+	err = bnnc.addRepPort(pod, dpuCD, podInterfaceInfo, getter)
+	if err != nil {
+		klog.Errorf("Failed to add rep port for %s, %v. retrying", podDesc, err)
+	}
+	return err
+}
+
+func dpuConnectionDetailChanged(oldDPUCD, newDPUCD *util.DPUConnectionDetails) bool {
+	if oldDPUCD == nil && newDPUCD == nil {
+		return false
+	}
+	if (oldDPUCD != nil && newDPUCD == nil) || (oldDPUCD == nil && newDPUCD != nil) {
+		return true
+	}
+	if oldDPUCD.PfId != newDPUCD.PfId ||
+		oldDPUCD.VfId != newDPUCD.VfId || oldDPUCD.SandboxId != newDPUCD.SandboxId {
+		return true
+	}
+	return false
+}
+
 // watchPodsDPU watch updates for pod dpu annotations
 func (bnnc *BaseNodeNetworkController) watchPodsDPU() error {
-	var retryPods sync.Map
-	// servedPods tracks the pods that got a VF
-	var servedPods sync.Map
-
 	clientSet := cni.NewClientSet(bnnc.client, corev1listers.NewPodLister(bnnc.watchFactory.LocalPodInformer().GetIndexer()))
 
-	// Support default network for now
-	nadName := types.DefaultNetworkName
 	netName := types.DefaultNetworkName
 	_, err := bnnc.watchFactory.AddPodHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			pod := obj.(*kapi.Pod)
-			klog.Infof("Add for Pod: %s/%s", pod.ObjectMeta.GetNamespace(), pod.ObjectMeta.GetName())
+			klog.V(5).Infof("Add for Pod: %s/%s for network %s", pod.Namespace, pod.Name, netName)
 			if util.PodWantsHostNetwork(pod) || pod.Status.Phase == kapi.PodRunning {
 				return
 			}
-			if util.PodScheduled(pod) {
-				// Is this pod created on same node as the DPU
-				if bnnc.name != pod.Spec.NodeName {
```

Comment:
Why was this code removed? Isn't it important to check if the pod is scheduled and that the pod's spec matches the node name?

Reply:
I see the second part is done in `podReadyToAddDPU()` now.

Reply:
But why is `PodScheduled()` no longer needed?

Reply:
PodScheduled() checks if pod.Spec.NodeName is not "". if bnnc.name is pod.Spec.NodeName, then it is not "".

Reply:
I see, thanks for the explanation, I guess we don't want to bail out early and just continue to run the code below to store the pod in `podNadCache` for the `UpdateFunc` case. I would like to see the cache to have some state storage instead of relying on existence checks on the maps. It makes following the cases a bit difficult. WDYT?

Reply:
but we need the list of NAD names on this network to addDPUPod for each NAD.

Reply:
Maybe what I am saying makes more sense if we consider combining the 2 cache maps. But if we don't plan to combine the 2 cache maps. Then what you have right now is no issue. It is just an suggestion to improve readability.

Reply:
even we merge the two cache, I'd prefer not move the PodScheduled check early in pod update handler, as we'd need to delete the VFRep if it has been configured if PodScheduled changed to false now. I am not sure if that will every happen, but logically it is correct to do so.